### PR TITLE
refactor: `Error` type

### PR DESCRIPTION
**Pull request by hal3e**
closes: https://github.com/FuelLabs/fuels-rs/issues/1236

All the suggestions from the issue were incorporated. In addition, error messages were changed to follow the guidelines from [rust-api-guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html?highlight=lowercase#error-types-are-meaningful-and-well-behaved-c-good-err) i.e. `The error message given by the Display representation of an error type should be lowercase without trailing punctuation, and typically concise.`

BREAKING CHANGE: the `fuels-rs` `Error` type is changed together with the error messages.

### Checklist
- [x] I have linked to any relevant issues.
- [x] I have updated the documentation.
- [x] I have added necessary labels.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: packages/fuels-accounts/src/predicate.rs [hunk 1]
```diff
@@ -98,8 +98,11 @@ impl ViewOnlyAccount for Predicate {
         self.address()
     }
 
-    fn try_provider(&self) -> AccountResult<&Provider> {
-        self.provider.as_ref().ok_or(AccountError::no_provider())
+    fn try_provider(&self) -> Result<&Provider> {
+        self.provider.as_ref().ok_or(error!(
+            Other,
+            "no provider available. Make sure to use `set_provider`"
```
Comment (MujkicA) [inline packages/fuels-accounts/src/predicate.rs:104]: We have a copy of this message in 3 places. Can we have it as a helper on `ViewOnlyAccount`

File: docs/src/calling-contracts/call-params.md [hunk 2]
```diff
@@ -28,7 +28,7 @@ Then, in Rust, after setting up and deploying the above contract, you can config
 <!-- payable:example:start -->
 `call_params` returns a result to ensure you don't forward assets to a contract method that isn't payable.
 <!-- payable:example:end -->
-In the following example, we try to forward an amount of 100 of the base asset to `non_payable`. As its name suggests, `non_payable` isn't annotated with `#[payable]` in the contract code. Passing `CallParameters` with an amount other than 0 leads to an `InvalidCallParameters` error:
+In the following example, we try to forward an amount of 100 of the base asset to `non_payable`. As its name suggests, `non_payable` isn't annotated with `#[payable]` in the contract code. Passing `CallParameters` with an amount other than `0` leads to an error:
```
[the remark this answers is not in the document]
Comment (hal3e) [reply docs/src/calling-contracts/call-params.md:31]: I do not think so that is why I removed the error name.

**Review by segfault-magnet — APPROVED** [review]
Great work. 

I'd like to raise just one issue while we're on the subject of errors. Sometimes I feel like we should add more context to our errors.

Due to the use of `?` in the sdk we just bubble up what are essentially low-level errors.

Failure to decode a contract response will be reported something like:
"Expected 8 B got 0" (off the  top of my head)

"Not enough coins to fit the target" 
without explaining that we were trying, for example, to cover gas costs  or cover the users asset requirements. 

Context can be added either by having every "layer" have its own errors or by having something akin to `anyhow`'s `Context` where we can add more context in the form of `String`s. 

Ultimately resulting in:
Failed to call `get_large_state(). Cause: Other, Not enough coins to fit the target
Context:
0. calling `get_large_state`
1. preparing transaction
2. adding coins to fit gas costs

Striking a balance between the verbosity of a backtrace and the terseness of 'Not enough coins to fit the target'. 

Since most of the context entries would not be something users would like to match on, maybe an option similar to `anyhow`'s `Context` would be best? 

Would mean wrapping the sdk `Error` in a struct and exposing the cause with a getter:
```rust
let call_result = contract_instance.methods()...;
if let Error(err) = call_result {
  let cause: Cause = err.cause();
  // match on cause if needed
  // Otherwise just print and get the cause and some context
  eprintln!("{}", err);
} 
```

File: packages/fuels-test-helpers/src/fuel_bin_service.rs
[the remark this answers is not in the document]
Comment (hal3e) [reply packages/fuels-test-helpers/src/fuel_bin_service.rs:1]: I would not use `InfrastructureError` anymore as it is very generic and the user can not do much at runtime if he encounters this kind of error.

**Review by segfault-magnet — APPROVED** [review]

**Review by Salka1988 — APPROVED** [review]

**Review by MujkicA — APPROVED** [review]

**Pull request merged**